### ISSUE TITLE
Unify gradient-based attacks

### ISSUE THREAD

**Issue by wielandbrendel** [bethgelab/foolbox#139]
Currently Foolbox features GradientSignAttack, GradientAttack, IterativeGradientAttack, IterativeGradientSignAttack. #138 proposes two more GradientAttacks using an additional Momentum term. Carlini & Wagner is basically just another gradient-based attack and uses Adam as well as special objective function.

In total this amounts to quite a bit of code replication that could be avoided. What I have in mind is a general GradientBaseAttack that takes as input:

- The Lp-norm to be optimized (e.g. L2 or Linfinity)
- The optimizer (plain gradient descent (GD), nesterov momentum, adam)
- The maximum number of iterations
- The loss function (e.g. cross-entropy or logits)

In this case the aforementioned attacks would just be shortcuts to

- GradientSignAttack = GradientBaseAttack(Linfinity, GD, 1)
- IterativeGradientSignAttack = GradientBaseAttack(Linfinity, GD, 1000)
- IterativeMomentumGradientSignAttack = GradientBaseAttack(Linfinity, momentum, 1000)
- GradientAttack = GradientBaseAttack(L2, GD, 1)

... and so on. In addition we would get many new attacks for free, in particular for different Lp metrics and optimizers. At the beginning it suffices to only implement L2 and Linfinity, but this modularization would give us the possibility to generalise to arbitrary Lp-norms quickly.

**Comment by jonasrauber** [comment]
I agree that some of the attacks can be unified like this, but I am not sure how well the Carlini & Wagner attack (and possibly other future attacks) will fit into this structure. CW basically performs an additional outer loop that optimizes a loss weighting hyper parameter. Extreme norms such as L0 and Linf also often require special treatment. But yes, unifying most of the gradient-based attacks would already be a great improvement and is absolutely doable.

**Comment by jonasrauber** [comment]
I've done this with the release of v1.2.0.
